In [ ]:
 # wab 2024.10.15 added in uniforms, high school, college, college stats

In [1]:
import pandas as pd

### IMPORT pro_ref_individual CSVs sourced and combine

In [4]:
file_1= "../Data_Artifacts/pro_ref_individ_raw.csv"
file_2= "../Data_Artifacts/pro_ref_individ_missing_raw.csv"

df_1 = pd.read_csv(file_1)
df_2 = pd.read_csv(file_2)

# file_2_fails = "../Data_Artifacts/pro_ref_individ_missing_fails_raw.csv"
# individ_missing_df = pd.read_csv(file_2_fails)

display(df_1.iloc[:,:10].head())
display(df_2.iloc[:,:10].head())
display(df_1.iloc[:,10:20].head())
display(df_2.iloc[:,10:20].head())
display(df_1.iloc[:,20:].head())
display(df_2.iloc[:,20:].head())

# display(individ_missing_df.head())

name throws  hall_of_fame                 college high_school   
0  Steve Bartkowski  Right        8842.0              California     Buchser  \
1   Steve Beuerlein  Right        8842.0              Notre Dame     Servite   
2    Archie Manning  Right        8842.0             Mississippi        Drew   
3        Brian Sipe  Right        8842.0  Grossmont College (CA)   Grossmont   
4       Andrew Luck  Right        8842.0                Stanford   Stratford   

   weighted_career_av                                      college_stats   
0               100.0  https://www.sports-reference.com/cfb/players/s...  \
1               100.0  https://www.sports-reference.com/cfb/players/s...   
2               100.0  https://www.sports-reference.com/cfb/players/a...   
3               100.0                                                NaN   
4               100.0  https://www.sports-reference.com/cfb/players/a...   

            draft_team  draft_round  draft_pick  
0      Atlanta Falcons          1.0         1.0  
1  Los Angeles Raiders          4.0       110.0  
2   New Orleans Saints          1.0         2.0  
3     Cleveland Browns         13.0       330.0  
4   Indianapolis Colts          1.0         1.0

name throws  hall_of_fame        college   
0           Tom Brady  Right        2000.0       Michigan  \
1          Drew Brees  Right        8842.0         Purdue   
2      Peyton Manning  Right        2000.0      Tennessee   
3         Brett Favre  Right        1990.0  Southern Miss   
4  Ben Roethlisberger  Right        8842.0     Miami (OH)   

             high_school  weighted_career_av   
0         Junipero Serra                 100  \
1               Westlake                 100   
2         Isidore Newman                 100   
3  Hancock North Central                 100   
4                Findlay                 100   

                                       college_stats            draft_team   
0  https://www.sports-reference.com/cfb/players/t...  New England Patriots  \
1  https://www.sports-reference.com/cfb/players/d...    San Diego Chargers   
2  https://www.sports-reference.com/cfb/players/p...    Indianapolis Colts   
3  https://www.sports-reference.com/cfb/players/b...       Atlanta Falcons   
4  https://www.sports-reference.com/cfb/players/b...   Pittsburgh Steelers   

   draft_round  draft_pick  
0          6.0       199.0  
1          2.0        32.0  
2          1.0         1.0  
3          2.0        33.0  
4          1.0        11.0

draft_year  games_played approximate_value qb_record   
0      1975.0         129.0                84   59-68-0  \
1      1987.0         147.0                85   47-55-0   
2      1971.0         151.0                94  35-101-3   
3      1972.0         125.0                87   57-55-0   
4      2012.0          86.0                80   53-33-0   

   completion_percentage  passing_yards  yards_per_attempt  touchdowns   
0                   55.9        24124.0                7.0       156.0  \
1                   56.9        24046.0                7.2       147.0   
2                   55.2        23911.0                6.6       125.0   
3                   56.5        23713.0                6.9       154.0   
4                   60.8        23671.0                7.2       171.0   

   interceptions  fantasy_points  
0          144.0          1329.0  
1          112.0          1333.7  
2          173.0          1364.6  
3          149.0          1347.6  
4           83.0          1686.2

draft_year  games_played approximate_value  qb_record   
0      2000.0           335               326   251-82-0  \
1      2001.0           287               277  172-114-0   
2      1998.0           266               271   186-79-0   
3      1991.0           302               259  186-112-0   
4      2004.0           249               208   165-81-1   

   completion_percentage  passing_yards  yards_per_attempt  touchdowns   
0                   64.3        89214.0                7.4       649.0  \
1                   67.7        80358.0                7.6       571.0   
2                   65.3        71940.0                7.7       539.0   
3                   62.0        71838.0                7.1       508.0   
4                   64.4        64088.0                7.6       418.0   

   interceptions  fantasy_points  
0          212.0          5943.4  
1          243.0          5185.0  
2          251.0          4686.3  
3          336.0          4371.2  
4          211.0          4019.5

all_star   
0                ['2x Pro Bowl']  \
1                ['1x Pro Bowl']   
2                ['2x Pro Bowl']   
3  ['1x Pro Bowl', '1x All-Pro']   
4                ['4x Pro Bowl']   

                                            uniforms           when_born   
0  ['Atlanta Falcons 1975-1985', 'Los Angeles Ram...   November 12, 1952  \
1  ['Los Angeles Raiders 1988-1989', 'Dallas Cowb...       March 7, 1965   
2  ['New Orleans Saints 1971-1982', 'Houston Oile...        May 19, 1949   
3                     ['Cleveland Browns 1974-1983']      August 8, 1949   
4                   ['Indianapolis Colts 2012-2018']  September 12, 1989   

        where_born height  weight  pass_yards     TD  superbowls  retire_year   
0   Des Moines, IA    6-4   216.0     24124.0  156.0         0.0       1986.0  \
1  Los Angeles, CA    6-3   220.0     24046.0  147.0         1.0       2003.0   
2    Cleveland, MS    6-3   212.0     23911.0  125.0         0.0       1984.0   
3    San Diego, CA    6-1   195.0     23713.0  154.0         0.0       1983.0   
4   Washington, DC    6-4   240.0     23671.0  171.0         0.0       2018.0   

         nickname  
0  Peachtree Bart  
1             NaN  
2             NaN  
3             NaN  
4             NaN

all_star   
0  ['15x Pro Bowl', '3x All-Pro']  \
1  ['13x Pro Bowl', '1x All-Pro']   
2  ['14x Pro Bowl', '7x All-Pro']   
3  ['11x Pro Bowl', '3x All-Pro']   
4                 ['6x Pro Bowl']   

                                            uniforms         when_born   
0  ['New England Patriots 2000-2019', 'Tampa Bay ...    August 3, 1977  \
1  ['San Diego Chargers 2001-2005', 'New Orleans ...  January 15, 1979   
2  ['Indianapolis Colts 1998-2010', 'Denver Bronc...    March 24, 1976   
3  ['Atlanta Falcons 1991', 'Green Bay Packers 19...  October 10, 1969   
4                  ['Pittsburgh Steelers 2004-2021']     March 2, 1982   

        where_born height  weight  pass_yards     TD  superbowls  retire_year   
0    San Mateo, CA    6-4     225     89214.0  649.0         7.0       2022.0  \
1       Austin, TX    6-0     209     80358.0  571.0         1.0       2020.0   
2  New Orleans, LA    6-5     230     71940.0  539.0         2.0       2015.0   
3     Gulfport, MS    6-2     222     71838.0  508.0         1.0       2010.0   
4         Lima, OH    6-5     240     64088.0  418.0         2.0       2021.0   

                                            nickname  
0  TB12, Tom Terrific, Touchdown Tom, GOAT, The P...  
1                                                NaN  
2                  The Sheriff, The Caveman, Peydirt  
3                Country, Gunslinger, Buckhead Brett  
4                                            Big Ben

### Concat Dataframes on axis = 1

In [9]:
concat_df = pd.concat([df_1, df_2], ignore_index=True)

display(concat_df.info())
display(concat_df.head())
display(f'# cols: {len(concat_df.columns)}; len df_1 rows: {len(df_1)}; df_2 rows: {len(df_2)}, concat_df rows: {len(concat_df)}')
display(concat_df.columns)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 582 entries, 0 to 581
Data columns (total 31 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   name                   582 non-null    object 
 1   throws                 431 non-null    object 
 2   hall_of_fame           286 non-null    float64
 3   college                563 non-null    object 
 4   high_school            500 non-null    object 
 5   weighted_career_av     465 non-null    float64
 6   college_stats          472 non-null    object 
 7   draft_team             582 non-null    object 
 8   draft_round            431 non-null    float64
 9   draft_pick             431 non-null    float64
 10  draft_year             425 non-null    float64
 11  games_played           475 non-null    float64
 12  approximate_value      396 non-null    object 
 13  qb_record              389 non-null    object 
 14  completion_percentage  399 non-null    float64
 15  passin

None

name throws  hall_of_fame                 college high_school   
0  Steve Bartkowski  Right        8842.0              California     Buchser  \
1   Steve Beuerlein  Right        8842.0              Notre Dame     Servite   
2    Archie Manning  Right        8842.0             Mississippi        Drew   
3        Brian Sipe  Right        8842.0  Grossmont College (CA)   Grossmont   
4       Andrew Luck  Right        8842.0                Stanford   Stratford   

   weighted_career_av                                      college_stats   
0               100.0  https://www.sports-reference.com/cfb/players/s...  \
1               100.0  https://www.sports-reference.com/cfb/players/s...   
2               100.0  https://www.sports-reference.com/cfb/players/a...   
3               100.0                                                NaN   
4               100.0  https://www.sports-reference.com/cfb/players/a...   

            draft_team  draft_round  draft_pick  ...   
0      Atlanta Falcons          1.0         1.0  ...  \
1  Los Angeles Raiders          4.0       110.0  ...   
2   New Orleans Saints          1.0         2.0  ...   
3     Cleveland Browns         13.0       330.0  ...   
4   Indianapolis Colts          1.0         1.0  ...   

                                            uniforms           when_born   
0  ['Atlanta Falcons 1975-1985', 'Los Angeles Ram...   November 12, 1952  \
1  ['Los Angeles Raiders 1988-1989', 'Dallas Cowb...       March 7, 1965   
2  ['New Orleans Saints 1971-1982', 'Houston Oile...        May 19, 1949   
3                     ['Cleveland Browns 1974-1983']      August 8, 1949   
4                   ['Indianapolis Colts 2012-2018']  September 12, 1989   

        where_born height  weight  pass_yards     TD  superbowls  retire_year   
0   Des Moines, IA    6-4   216.0     24124.0  156.0         0.0       1986.0  \
1  Los Angeles, CA    6-3   220.0     24046.0  147.0         1.0       2003.0   
2    Cleveland, MS    6-3   212.0     23911.0  125.0         0.0       1984.0   
3    San Diego, CA    6-1   195.0     23713.0  154.0         0.0       1983.0   
4   Washington, DC    6-4   240.0     23671.0  171.0         0.0       2018.0   

         nickname  
0  Peachtree Bart  
1             NaN  
2             NaN  
3             NaN  
4             NaN  

[5 rows x 31 columns]

'# cols: 31; len df_1 rows: 497; df_2 rows: 85, concat_df rows: 582'

Index(['name', 'throws', 'hall_of_fame', 'college', 'high_school',
       'weighted_career_av', 'college_stats', 'draft_team', 'draft_round',
       'draft_pick', 'draft_year', 'games_played', 'approximate_value',
       'qb_record', 'completion_percentage', 'passing_yards',
       'yards_per_attempt', 'touchdowns', 'interceptions', 'fantasy_points',
       'all_star', 'uniforms', 'when_born', 'where_born', 'height', 'weight',
       'pass_yards', 'TD', 'superbowls', 'retire_year', 'nickname'],
      dtype='object')

In [10]:
# reduce the number of columns:
cleaning_df = concat_df[[
    'name', 'hall_of_fame', 'all_star', 'superbowls',    
    'weighted_career_av', 'approximate_value', 'qb_record', 'games_played', 
    'pass_yards', 'TD', 'completion_percentage', 
    'yards_per_attempt', 'interceptions', 
    'draft_year', 'draft_team', 'draft_round', 'draft_pick', 
    'uniforms', 'high_school', 'college', 'college_stats',   # wab 2024.10.15 added in uniforms, high school, college, college stats
    'height', 'weight', 'throws'
]]

display(f'# cols: {len(cleaning_df.columns)}; cleaning_df rows: {len(cleaning_df)}')
display(concat_df.columns)

'# cols: 24; cleaning_df rows: 582'

Index(['name', 'throws', 'hall_of_fame', 'college', 'high_school',
       'weighted_career_av', 'college_stats', 'draft_team', 'draft_round',
       'draft_pick', 'draft_year', 'games_played', 'approximate_value',
       'qb_record', 'completion_percentage', 'passing_yards',
       'yards_per_attempt', 'touchdowns', 'interceptions', 'fantasy_points',
       'all_star', 'uniforms', 'when_born', 'where_born', 'height', 'weight',
       'pass_yards', 'TD', 'superbowls', 'retire_year', 'nickname'],
      dtype='object')

In [16]:
cleaning_df.rename(columns={
                    'name': 'player',
                    'hall_of_fame': 'hof_yr', #  # wab 2024.10.15 changed to hof_yr
                    'all_star': 'all_star',
                    'uniforms': 'teams',
                    'superbowls': 'SBs',
                    'weighted_career_av': 'wAV',
                    'approximate_value': 'AV',
                    'qb_record': 'record',
                    'games_played': 'games',
                    'pass_yards': 'pass_yards',
                    'TD': 'TD',
                    'completion_percentage': 'comp_%',
                    'yards_per_attempt': 'YPA',
                    'interceptions': 'int',
                    'draft_year': 'draft_yr',    # wab 2024.10.15  changed to draft_yr from draft
                    'draft_team': 'draft_team',  # wab 2024.10.15  changed to draft_team from team
                    'draft_round': 'rd',
                    'draft_pick': 'pick',
                    'height': 'ht',
                    'weight': 'wt',
                    'throws': 'hand'
                }, inplace=True)

display(cleaning_df.columns)
display(cleaning_df.iloc[:,:10].head())
display(cleaning_df.iloc[:,10:].head())

C:\Users\black\AppData\Local\Temp\ipykernel_16328\1841485237.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaning_df.rename(columns={


Index(['player', 'hof_yr', 'all_star', 'SBs', 'wAV', 'AV', 'record', 'games',
       'pass_yards', 'TD', 'comp_%', 'YPA', 'int', 'draft_yr', 'draft_team',
       'rd', 'pick', 'teams', 'high_school', 'college', 'college_stats', 'ht',
       'wt', 'hand', 'wins', 'loss'],
      dtype='object')

player  hof_yr                       all_star  SBs    wAV  AV   
0  Steve Bartkowski  8842.0                ['2x Pro Bowl']  0.0  100.0  84  \
1   Steve Beuerlein  8842.0                ['1x Pro Bowl']  1.0  100.0  85   
2    Archie Manning  8842.0                ['2x Pro Bowl']  0.0  100.0  94   
3        Brian Sipe  8842.0  ['1x Pro Bowl', '1x All-Pro']  0.0  100.0  87   
4       Andrew Luck  8842.0                ['4x Pro Bowl']  0.0  100.0  80   

     record  games  pass_yards     TD  
0   59-68-0  129.0     24124.0  156.0  
1   47-55-0  147.0     24046.0  147.0  
2  35-101-3  151.0     23911.0  125.0  
3   57-55-0  125.0     23713.0  154.0  
4   53-33-0   86.0     23671.0  171.0

comp_%  YPA    int  draft_yr           draft_team    rd   pick   
0    55.9  7.0  144.0    1975.0      Atlanta Falcons   1.0    1.0  \
1    56.9  7.2  112.0    1987.0  Los Angeles Raiders   4.0  110.0   
2    55.2  6.6  173.0    1971.0   New Orleans Saints   1.0    2.0   
3    56.5  6.9  149.0    1972.0     Cleveland Browns  13.0  330.0   
4    60.8  7.2   83.0    2012.0   Indianapolis Colts   1.0    1.0   

                                               teams high_school   
0  ['Atlanta Falcons 1975-1985', 'Los Angeles Ram...     Buchser  \
1  ['Los Angeles Raiders 1988-1989', 'Dallas Cowb...     Servite   
2  ['New Orleans Saints 1971-1982', 'Houston Oile...        Drew   
3                     ['Cleveland Browns 1974-1983']   Grossmont   
4                   ['Indianapolis Colts 2012-2018']   Stratford   

                  college                                      college_stats   
0              California  https://www.sports-reference.com/cfb/players/s...  \
1              Notre Dame  https://www.sports-reference.com/cfb/players/s...   
2             Mississippi  https://www.sports-reference.com/cfb/players/a...   
3  Grossmont College (CA)                                                NaN   
4                Stanford  https://www.sports-reference.com/cfb/players/a...   

    ht     wt   hand  wins   loss  
0  6-4  216.0  Right  59.0   68.0  
1  6-3  220.0  Right  47.0   55.0  
2  6-3  212.0  Right  35.0  101.0  
3  6-1  195.0  Right  57.0   55.0  
4  6-4  240.0  Right  53.0   33.0

In [17]:
# split out the record column into two new columns: wins/losses:
import numpy as np

def split_record(record):
    if pd.isna(record):
        return pd.Series([np.nan, np.nan])
    else:
        parts = record.split('-')
        if len(parts) >= 2 and parts[0].isdigit() and parts[1].isdigit():
            return pd.Series([int(parts[0]), int(parts[1])])
        else:
            return pd.Series([np.nan, np.nan])

# Apply the function to the record column
cleaning_df[['wins', 'loss']] = cleaning_df['record'].apply(split_record)
cleaning_df.head()


C:\Users\black\AppData\Local\Temp\ipykernel_16328\4262382936.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaning_df[['wins', 'loss']] = cleaning_df['record'].apply(split_record)


player  hof_yr                       all_star  SBs    wAV  AV   
0  Steve Bartkowski  8842.0                ['2x Pro Bowl']  0.0  100.0  84  \
1   Steve Beuerlein  8842.0                ['1x Pro Bowl']  1.0  100.0  85   
2    Archie Manning  8842.0                ['2x Pro Bowl']  0.0  100.0  94   
3        Brian Sipe  8842.0  ['1x Pro Bowl', '1x All-Pro']  0.0  100.0  87   
4       Andrew Luck  8842.0                ['4x Pro Bowl']  0.0  100.0  80   

     record  games  pass_yards     TD  ...   pick   
0   59-68-0  129.0     24124.0  156.0  ...    1.0  \
1   47-55-0  147.0     24046.0  147.0  ...  110.0   
2  35-101-3  151.0     23911.0  125.0  ...    2.0   
3   57-55-0  125.0     23713.0  154.0  ...  330.0   
4   53-33-0   86.0     23671.0  171.0  ...    1.0   

                                               teams  high_school   
0  ['Atlanta Falcons 1975-1985', 'Los Angeles Ram...      Buchser  \
1  ['Los Angeles Raiders 1988-1989', 'Dallas Cowb...      Servite   
2  ['New Orleans Saints 1971-1982', 'Houston Oile...         Drew   
3                     ['Cleveland Browns 1974-1983']    Grossmont   
4                   ['Indianapolis Colts 2012-2018']    Stratford   

                  college                                      college_stats   
0              California  https://www.sports-reference.com/cfb/players/s...  \
1              Notre Dame  https://www.sports-reference.com/cfb/players/s...   
2             Mississippi  https://www.sports-reference.com/cfb/players/a...   
3  Grossmont College (CA)                                                NaN   
4                Stanford  https://www.sports-reference.com/cfb/players/a...   

    ht     wt   hand  wins   loss  
0  6-4  216.0  Right  59.0   68.0  
1  6-3  220.0  Right  47.0   55.0  
2  6-3  212.0  Right  35.0  101.0  
3  6-1  195.0  Right  57.0   55.0  
4  6-4  240.0  Right  53.0   33.0  

[5 rows x 26 columns]

### NEW DATAFRAME:  num_cleaning_df

In [18]:
# # Fill NaN values with 0 or another appropriate value
import numpy as np
num_cleaning_df = cleaning_df.copy()

# Replace non values in 'hof' and 'team' columns
num_cleaning_df['hof_yr'] = num_cleaning_df['hof_yr'].replace([8842, 4423], np.nan)
num_cleaning_df['draft_yr'] = num_cleaning_df['draft_yr'].replace('Draft History', 'Undrafted')

# Calculate win percentage in new column for win_% with our new win/loss columns:
num_cleaning_df['win_%'] = round(num_cleaning_df['wins'] / (num_cleaning_df['wins'] + num_cleaning_df['loss']) * 100, 2)

# Move 'win_%' column to the position right after 'games' column
win_pct_pop = num_cleaning_df.pop('win_%')
games_index = num_cleaning_df.columns.get_loc('games')
num_cleaning_df.insert(games_index + 1, 'win_%', win_pct_pop)

display(num_cleaning_df.head())
display(num_cleaning_df.tail())
num_cleaning_df.columns

player  hof_yr                       all_star  SBs    wAV  AV   
0  Steve Bartkowski     NaN                ['2x Pro Bowl']  0.0  100.0  84  \
1   Steve Beuerlein     NaN                ['1x Pro Bowl']  1.0  100.0  85   
2    Archie Manning     NaN                ['2x Pro Bowl']  0.0  100.0  94   
3        Brian Sipe     NaN  ['1x Pro Bowl', '1x All-Pro']  0.0  100.0  87   
4       Andrew Luck     NaN                ['4x Pro Bowl']  0.0  100.0  80   

     record  games  win_%  pass_yards  ...   pick   
0   59-68-0  129.0  46.46     24124.0  ...    1.0  \
1   47-55-0  147.0  46.08     24046.0  ...  110.0   
2  35-101-3  151.0  25.74     23911.0  ...    2.0   
3   57-55-0  125.0  50.89     23713.0  ...  330.0   
4   53-33-0   86.0  61.63     23671.0  ...    1.0   

                                               teams  high_school   
0  ['Atlanta Falcons 1975-1985', 'Los Angeles Ram...      Buchser  \
1  ['Los Angeles Raiders 1988-1989', 'Dallas Cowb...      Servite   
2  ['New Orleans Saints 1971-1982', 'Houston Oile...         Drew   
3                     ['Cleveland Browns 1974-1983']    Grossmont   
4                   ['Indianapolis Colts 2012-2018']    Stratford   

                  college                                      college_stats   
0              California  https://www.sports-reference.com/cfb/players/s...  \
1              Notre Dame  https://www.sports-reference.com/cfb/players/s...   
2             Mississippi  https://www.sports-reference.com/cfb/players/a...   
3  Grossmont College (CA)                                                NaN   
4                Stanford  https://www.sports-reference.com/cfb/players/a...   

    ht     wt   hand  wins   loss  
0  6-4  216.0  Right  59.0   68.0  
1  6-3  220.0  Right  47.0   55.0  
2  6-3  212.0  Right  35.0  101.0  
3  6-1  195.0  Right  57.0   55.0  
4  6-4  240.0  Right  53.0   33.0  

[5 rows x 27 columns]

player  hof_yr                       all_star  SBs    wAV   AV   
577        Bob Griese  1971.0  ['8x Pro Bowl', '2x All-Pro']  2.0  100.0  138  \
578      Tommy Kramer     NaN                ['1x Pro Bowl']  0.0  100.0   84   
579        Bart Starr  1966.0  ['4x Pro Bowl', '1x All-Pro']  2.0  100.0  115   
580   Charley Johnson     NaN                             []  1.0  100.0  NaN   
581  Daunte Culpepper     NaN                ['3x Pro Bowl']  0.0  100.0   98   

      record  games  win_%  pass_yards  ...   pick   
577  92-56-3  161.0  62.16     25092.0  ...    4.0  \
578  54-56-0  129.0  49.09     24777.0  ...   27.0   
579  94-57-6  196.0  62.25     24718.0  ...  200.0   
580      NaN  133.0    NaN         NaN  ...   17.0   
581  41-59-0  105.0  41.00     24153.0  ...   11.0   

                                                 teams    high_school   
577                       ['Miami Dolphins 1967-1980']      Rex Mundi  \
578  ['Minnesota Vikings 1977-1989', 'New Orleans S...  Robert E. Lee   
579  ['Green Bay Packers 1956', 'Green Bay Packers ...  Sidney Lanier   
580  ['Pittsburgh Steelers 1994-1998', 'Philadelphi...          Cajon   
581  ['Minnesota Vikings 1999', 'Minnesota Vikings ...       Vanguard   

             college                                      college_stats   ht   
577           Purdue  https://www.sports-reference.com/cfb/players/b...  6-1  \
578             Rice  https://www.sports-reference.com/cfb/players/t...  6-2   
579          Alabama  https://www.sports-reference.com/cfb/players/b...  6-1   
580         Colorado  https://www.sports-reference.com/cfb/players/c...  6-0   
581  Central Florida  https://www.sports-reference.com/cfb/players/d...  6-4   

        wt   hand  wins  loss  
577  190.0  Right  92.0  56.0  
578  200.0  Right  54.0  56.0  
579  197.0  Right  94.0  57.0  
580  200.0    NaN   NaN   NaN  
581  260.0  Right  41.0  59.0  

[5 rows x 27 columns]

Index(['player', 'hof_yr', 'all_star', 'SBs', 'wAV', 'AV', 'record', 'games',
       'win_%', 'pass_yards', 'TD', 'comp_%', 'YPA', 'int', 'draft_yr',
       'draft_team', 'rd', 'pick', 'teams', 'high_school', 'college',
       'college_stats', 'ht', 'wt', 'hand', 'wins', 'loss'],
      dtype='object')

In [19]:
# convert cols to int
cols_to_int = ['wins', 'loss', 'hof_yr', 'SBs', 'games', 'TD', 'draft_yr', 'rd', 'pick']

# Replace blank values with NaN  
# REGEX: r: raw string, ^: beginning of string; \s: whitespace; *: zero or more occurances of preceding element; $: end of string
num_cleaning_df[cols_to_int] = num_cleaning_df[cols_to_int].replace(r'^\s*$', np.nan, regex=True)

# Convert columns to nullable integer type
# To leave values as NaN & still convert columns to int64, use pd.Int64Dtype() (allows for nullable integer columns)
num_cleaning_df[cols_to_int] = num_cleaning_df[cols_to_int].astype(pd.Int64Dtype())

display(num_cleaning_df.head())
display(num_cleaning_df.tail())
num_cleaning_df.columns


player  hof_yr                       all_star  SBs    wAV  AV   
0  Steve Bartkowski    <NA>                ['2x Pro Bowl']    0  100.0  84  \
1   Steve Beuerlein    <NA>                ['1x Pro Bowl']    1  100.0  85   
2    Archie Manning    <NA>                ['2x Pro Bowl']    0  100.0  94   
3        Brian Sipe    <NA>  ['1x Pro Bowl', '1x All-Pro']    0  100.0  87   
4       Andrew Luck    <NA>                ['4x Pro Bowl']    0  100.0  80   

     record  games  win_%  pass_yards  ...  pick   
0   59-68-0    129  46.46     24124.0  ...     1  \
1   47-55-0    147  46.08     24046.0  ...   110   
2  35-101-3    151  25.74     23911.0  ...     2   
3   57-55-0    125  50.89     23713.0  ...   330   
4   53-33-0     86  61.63     23671.0  ...     1   

                                               teams  high_school   
0  ['Atlanta Falcons 1975-1985', 'Los Angeles Ram...      Buchser  \
1  ['Los Angeles Raiders 1988-1989', 'Dallas Cowb...      Servite   
2  ['New Orleans Saints 1971-1982', 'Houston Oile...         Drew   
3                     ['Cleveland Browns 1974-1983']    Grossmont   
4                   ['Indianapolis Colts 2012-2018']    Stratford   

                  college                                      college_stats   
0              California  https://www.sports-reference.com/cfb/players/s...  \
1              Notre Dame  https://www.sports-reference.com/cfb/players/s...   
2             Mississippi  https://www.sports-reference.com/cfb/players/a...   
3  Grossmont College (CA)                                                NaN   
4                Stanford  https://www.sports-reference.com/cfb/players/a...   

    ht     wt   hand wins loss  
0  6-4  216.0  Right   59   68  
1  6-3  220.0  Right   47   55  
2  6-3  212.0  Right   35  101  
3  6-1  195.0  Right   57   55  
4  6-4  240.0  Right   53   33  

[5 rows x 27 columns]

player  hof_yr                       all_star  SBs    wAV   AV   
577        Bob Griese    1971  ['8x Pro Bowl', '2x All-Pro']    2  100.0  138  \
578      Tommy Kramer    <NA>                ['1x Pro Bowl']    0  100.0   84   
579        Bart Starr    1966  ['4x Pro Bowl', '1x All-Pro']    2  100.0  115   
580   Charley Johnson    <NA>                             []    1  100.0  NaN   
581  Daunte Culpepper    <NA>                ['3x Pro Bowl']    0  100.0   98   

      record  games  win_%  pass_yards  ...  pick   
577  92-56-3    161  62.16     25092.0  ...     4  \
578  54-56-0    129  49.09     24777.0  ...    27   
579  94-57-6    196  62.25     24718.0  ...   200   
580      NaN    133    NaN         NaN  ...    17   
581  41-59-0    105  41.00     24153.0  ...    11   

                                                 teams    high_school   
577                       ['Miami Dolphins 1967-1980']      Rex Mundi  \
578  ['Minnesota Vikings 1977-1989', 'New Orleans S...  Robert E. Lee   
579  ['Green Bay Packers 1956', 'Green Bay Packers ...  Sidney Lanier   
580  ['Pittsburgh Steelers 1994-1998', 'Philadelphi...          Cajon   
581  ['Minnesota Vikings 1999', 'Minnesota Vikings ...       Vanguard   

             college                                      college_stats   ht   
577           Purdue  https://www.sports-reference.com/cfb/players/b...  6-1  \
578             Rice  https://www.sports-reference.com/cfb/players/t...  6-2   
579          Alabama  https://www.sports-reference.com/cfb/players/b...  6-1   
580         Colorado  https://www.sports-reference.com/cfb/players/c...  6-0   
581  Central Florida  https://www.sports-reference.com/cfb/players/d...  6-4   

        wt   hand  wins  loss  
577  190.0  Right    92    56  
578  200.0  Right    54    56  
579  197.0  Right    94    57  
580  200.0    NaN  <NA>  <NA>  
581  260.0  Right    41    59  

[5 rows x 27 columns]

Index(['player', 'hof_yr', 'all_star', 'SBs', 'wAV', 'AV', 'record', 'games',
       'win_%', 'pass_yards', 'TD', 'comp_%', 'YPA', 'int', 'draft_yr',
       'draft_team', 'rd', 'pick', 'teams', 'high_school', 'college',
       'college_stats', 'ht', 'wt', 'hand', 'wins', 'loss'],
      dtype='object')

### EXPORT CSV    this time pro_ref_individ.csv

In [20]:
#Export csv
rel_path = "../Data_Artifacts/pro_ref_individ_merged.csv"
num_cleaning_df.to_csv(rel_path, index=False)

### ** ALL REMAINING CELLS: NOT USED **

In [ ]:
import pandas as pd

# reload pro ref individuals csv
rel_path = "../Data_Artifacts/pro_ref_individ.csv"
_df = pd.read_csv(rel_path)

In [ ]:
#  to_do : 
# look at the original QB Data Dataframe (with its data, and look at the values of
#  these quarterbacks) qb_name_df

#  Filter the dataframe by the list of quarterback fails and be sure no significant names were lost.
qb_fails_df = pd.DataFrame(quarterback_fails)
display(f'{len(qb_fails_df)}')
display(qb_fails_df.head())
display(qb_fails_df.info())

# display(quarterback_dict)
# display(quarterback_dict.info())

In [ ]:
import pprint
# Print the combined data for verification
print("\quarterback_dict:")
pp = pprint.PrettyPrinter(indent=4)
pp.pprint(quarterback_dict)

### SAMPLE DATA FOR INDIVIDUAL PLAYER

<div id="info" class="players">
	<div id="meta">
<div  class="nothumb">
	<h1>
		<span>Steve Bartkowski</span>
			</h1>
<p>
  <strong>
    Steven Joseph Bartkowski
    &nbsp;(<span>Peachtree Bart</span>)
  </strong>
</p>
<p>
	<strong>Position</strong>: QB
			<strong>Throws:</strong>
		Right
</p>
<p><span>6-4</span>,&nbsp;<span>216lb</span>&nbsp;(193cm,&nbsp;97kg) </p>
<p>
  <strong>Born:</strong> 
    <span" id="necro-birth" data-birth="1952-11-12">
		<a href="/friv/birthdays.cgi?month=11&day=12">November 12</a>,&nbsp;1952
    </span>
  <span>
    in&nbsp;Des Moines,&nbsp;<a href="/friv/birthplaces.cgi?country=USA&state=IA">IA</a></span>
</p>
<p><strong>College</strong>: 
		<a href="/schools/california/">California</a>
		&nbsp;(<a href="https://www.sports-reference.com/cfb/players/steve-bartkowski-1.html">College Stats</a>)
<p><strong><a href="https://www.pro-football-reference.com/about/approximate_value.htm">Weighted Career AV (100-95-...)</a>:</strong> 69 (703rd overall since 1960)
</p>
<p><strong>High School</strong>: 
		<a href="/schools/high_schools.cgi?id=93b907c3">Buchser</a> (<a href="/schools/high_schools.cgi?hs_state=CA">CA</a>)
</p>
<p><strong>Draft</strong>: <a href="/teams/atl/1975_draft.htm">Atlanta Falcons</a> in the 1st round (1st overall) of the <a href="/years/1975/draft.htm">1975 NFL Draft</a>.</p>
</p>
</div>
</div><!-- div#meta -->

<ul id="bling">
<li class="all_star"><a href="/probowl/">2x Pro Bowl</a></li><li class=""><a href="/awards/">1975 SN RoY</a></li>
</ul>	
	<div class="uni_holder pfr">
	<a href="/players/uniform.cgi?team=atl&number=10" class="poptip default" data-tip="Atlanta Falcons 1975-1985">
	</a>
	<a href="/players/uniform.cgi?team=ram&number=10" class="poptip" data-tip="Los Angeles Rams 1986">
	</a>
	<div class="counter opener" data-class="open" data-id="info" data-control-id="meta_more_button" data-type="hide_after">+1</div>
</div>
	<span id="bling-alt-text" style="display:none;"><!--
Recognition:
* 2x Pro Bowl
* 1975 SN RoY
Uniforms:
* Number 10 for Atlanta Falcons 1975-1985
* Number 10 for Los Angeles Rams 1986
--></span>		
</div>

<div class="stats_pullout">
<div><div>
<span><strong>SUMMARY</strong></span>
<p><strong>Career</strong></p>
</div></div>
<div class="p1">
<div class="p1">
<span class="poptip" data-tip="Games played"><strong>G</strong></span>
<p>129</p></div>

<div class="p1">
<span class="poptip" data-tip="Approximate Value is our attempt to attach a single number to every player-season since 1960.<br>See our About section for more details."><strong>AV</strong></span>
<p>84</p></div>
</div><div class="p1">
<div class="p2">
<span class="poptip" data-tip="Team record in games started by this QB (regular season)"><strong>QBrec</strong></span>
<p>59-68-0</p></div>

<div class="p1">
<span class="poptip" data-tip="Percentage of Passes Completed<br>Minimum 14 attempts per scheduled game to qualify as leader.<br />Minimum 1500 pass attempts to qualify as career leader."><strong>Cmp%</strong></span>
<p>55.9</p></div>

<div class="p1">
<span class="poptip" data-tip="Yards Gained by Passing<br>For teams, sack yardage is deducted from this total"><strong>Yds</strong></span>
<p>24124</p></div>

<div class="p2">
<span class="poptip" data-tip="Yards gained per pass attempt <br>Minimum 14 attempts per scheduled game to qualify as leader.<br>Minimum 1500 pass attempts to qualify as career leader."><strong>Y/A</strong></span>
<p>7.0</p></div>

<div class="p1">
<span class="poptip" data-tip="Passing Touchdowns"><strong>TD</strong></span>
<p>156</p></div>

<div class="p2">
<span class="poptip" data-tip="Interceptions thrown"><strong>Int</strong></span>
<p>144</p></div>

<div class="p2">
<span class="poptip" data-tip="<b>Fantasy points:</b><br />
								1 point per 25 yards passing<br />
								4 points per passing touchdown<br />
								-2 points per interception thrown<br />
								1 point per 10 yards rushing/receiving<br />
								6 points per TD<br />
								2 points per two-point conversion<br />
								-2 points per fumble lost (est. prior to 1994)"><strong>FantPt</strong></span>
<p>1329.0</p></div>
</div>
</div>
</div><!-- div#info -->

In [ ]:
# For FAQ section:

# <div class="section_content" id="div_faq">
#   <h3>How old is Steve Bartkowski?</h3>
#   <p>Steve Bartkowski is 71 years old.</p>
#   <h3>When was Steve Bartkowski born?</h3>
#   <p>Steve Bartkowski was born on November 12, 1952.</p>
#   <h3>Where was Steve Bartkowski born?</h3>
#   <p>Steve Bartkowski was born in Des Moines, IA.</p>
#   <h3>How tall is Steve Bartkowski?</h3>
#   <p>Steve Bartkowski is 6-4 (193 cm) tall.</p>
#   <h3>How much did Steve Bartkowski weigh when playing?</h3>
#   <p>Steve Bartkowski weighed 216 lbs (97 kg) when playing.</p>
#   <h3>Is Steve Bartkowski in the Hall of Fame?</h3>
#   <p>Steve Bartkowski has not been elected into the Hall of Fame.</p>
#   <h3>How many games did Steve Bartkowski play?</h3>
#   <p>Steve Bartkowski played 129 games over his career.</p>
#   <h3>How many passing yards did Steve Bartkowski have?</h3>
#   <p>Steve Bartkowski had 24,124 passing yards over his career.</p>
#   <h3>How many touchdowns did Steve Bartkowski have?</h3>
#   <p>Steve Bartkowski had 156 touchdowns over his career.</p>
#   <h3>How many Super Bowls has Steve Bartkowski won?</h3>
#   <p>Steve Bartkowski has never won a Super Bowl.</p>
#   <h3>When did Steve Bartkowski retire?</h3>
#   <p>Steve Bartkowski last played in 1986.</p>
#   <h3>What are Steve Bartkowski's nicknames?</h3>
#   <p>Peachtree Bart is a nickname for Steve Bartkowski.</p>